src/lib/yang/data.lua

```diff
@@ -0,0 +1,284 @@
+-- Use of this source code is governed by the Apache 2.0 license; see
+-- COPYING.
+module(..., package.seeall)
+
+local parse_string = require("lib.yang.parser").parse_string
+local schema = require("lib.yang.schema")
+local util = require("lib.yang.util")
+local value = require("lib.yang.value")
+
+-- FIXME:
+-- Parse inet:mac-address using ethernet:pton
+-- Parse inet:ipv4-address using ipv4:pton
+-- Parse inet:ipv6-address using ipv6:pton
+-- Parse inet:ipv4-prefix?
+-- Parse inet:ipv6-prefix?
+
+function data_grammar_from_schema(schema)
+   local handlers = {}
+   local function visit(node)
+      local handler = handlers[node.kind]
+      if handler then return handler(node) end
+      return {}
+   end
+   local function visit_body(node)
+      local ret = {}
+      for id,node in pairs(node.body) do
+         for keyword,node in pairs(visit(node)) do
+            assert(not ret[keyword], 'duplicate identifier: '..keyword)
+            ret[keyword] = node
+         end
+      end
+      return ret
+   end
+   function handlers.container(node)
+      if not node.presence then return visit_body(node) end
+      return {[node.id]={type='struct', members=visit_body(node)}}
+   end
+   handlers['leaf-list'] = function(node)
+      return {[node.id]={type='array', element_type=node.type}}
+   end
+   function handlers.list(node)
+      local keys = {}
+      for key in node.key:split(' +') do table.insert(keys, key) end
+      return {[node.id]={type='table', keys=keys, members=visit_body(node)}}
+   end
+   function handlers.leaf(node)
+      return {[node.id]={type='scalar', argument_type=node.type,
+                         default=node.default, mandatory=node.mandatory}}
+   end
+   return {type="struct", members=visit_body(schema)}
+end
+
+local function integer_type(min, max)
+   return function(str, k)
+      return util.tointeger(str, k, min, max)
+   end
+end
+
+-- FIXME :)
+local function range_validator(range, f) return f end
+local function length_validator(range, f) return f end
+local function pattern_validator(range, f) return f end
+local function bit_validator(range, f) return f end
+local function enum_validator(range, f) return f end
+
+local function value_parser(typ)
+   local prim = typ.primitive_type
+   local parse = assert(value.types[prim], prim).parse
```

Comment:
This error message could be more useful, currently it's just "Error in core/main.lua:26: enumeration".